core/src/main/kotlin/dev/hotwire/core/turbo/session/Session.kt

```diff
@@ -277,24 +296,76 @@ class Session(
      * Warning: This method is public so it can be used as a Javascript Interface.
      * You should never call this directly as it could lead to unintended behavior.
      *
+     * @param location The location of the failed visit.
      * @param visitIdentifier A unique identifier for the visit.
      * @param visitHasCachedSnapshot Whether the visit has a cached snapshot available.
      * @param statusCode The HTTP status code that caused the failure.
      */
     @JavascriptInterface
-    fun visitRequestFailedWithStatusCode(visitIdentifier: String, visitHasCachedSnapshot: Boolean, statusCode: Int) {
+    fun visitRequestFailedWithStatusCode(
+        location: String,
+        visitIdentifier: String,
+        visitHasCachedSnapshot: Boolean,
+        statusCode: Int
+    ) {
         val visitError = HttpError.from(statusCode)
 
         logEvent(
             "visitRequestFailedWithStatusCode",
+            "location" to location,
             "visitIdentifier" to visitIdentifier,
             "visitHasCachedSnapshot" to visitHasCachedSnapshot,
             "error" to visitError
         )
 
-        currentVisit?.let { visit ->
-            if (visitIdentifier == visit.identifier) {
-                callback { it.requestFailedWithError(visitHasCachedSnapshot, visitError) }
+        if (visitIdentifier == currentVisit?.identifier) {
+            callback { it.requestFailedWithError(visitHasCachedSnapshot, visitError) }
+        }
+    }
+
+    /**
+     * Called by Turbo bridge when a visit request fails with a non-HTTP status code, suggesting
+     * it may be the result of a cross-origin redirect visit. Determining a cross-origin redirect
+     * is not possible in javascript with the Fetch API due to CORS restrictions, so verify on
+     * the native side. Propose a cross-origin redirect visit if a redirect is found, otherwise
+     * fail the visit.
+     *
+     * Warning: This method is public so it can be used as a Javascript Interface.
+     * You should never call this directly as it could lead to unintended behavior.
+     *
+     * @param location The original visit location requested.
+     * @param visitIdentifier A unique identifier for the visit.
+     * @param visitHasCachedSnapshot Whether the visit has a cached snapshot available.
+     */
+    @JavascriptInterface
+    fun visitRequestFailedWithNonHttpStatusCode(
+        location: String,
+        visitIdentifier: String,
+        visitHasCachedSnapshot: Boolean
+    ) {
+        logEvent("visitRequestFailedWithNonHttpStatusCode",
+            "location" to location,
+            "visitIdentifier" to visitIdentifier,
+            "visitHasCachedSnapshot" to visitHasCachedSnapshot
+        )
+
+        activity.lifecycleScope.launch {
+            val result = httpRepository.fetch(location)
+
+            if (result != null && result.response.isSuccessful &&
+                result.redirectToLocation != null && result.redirectIsCrossOrigin) {
```

Comment:
`result.redirectToLocation != null` seems to be checked again here even though it's already part of `result.redirectIsCrossOrigin` value computation

Reply:
To improve this situation, I moved the redirect-specific properties under `result.redirect` and now we just have to check:
```kotlin
result.redirect?.isCrossOrigin == true
```

Reply:
Perfect!